Sources/SwiftFusion/Inference/ArrayBuffer+Vector.swift

```diff
@@ -13,7 +13,97 @@
 // limitations under the License.
 import PenguinStructures
 
+public struct NestedIndex<OuterIndex: Comparable, InnerIndex: Comparable> : Comparable {
+  fileprivate var outer: OuterIndex
+  fileprivate var inner: InnerIndex?
+
+  public static func <(lhs: Self, rhs: Self) -> Bool {
+    if lhs.outer < rhs.outer { return true }
+    if lhs.outer != rhs.outer { return false }
+    guard let l1 = lhs.inner else { return false }
+    guard let r1 = rhs.inner else { return true }
+    return l1 < r1
+  }
+
+  private init(outer: OuterIndex, inner: InnerIndex?) {
+    self.outer = outer
+    self.inner = inner
+  }
+  
+  internal init<OuterCollection: Collection, InnerCollection: Collection>(
+    firstValidIn c: OuterCollection, innerCollection: (OuterCollection.Element)->InnerCollection
+  )
+    where OuterCollection.Index == OuterIndex, InnerCollection.Index == InnerIndex
+  {
+    if let i = c.firstIndex(where: { !innerCollection($0).isEmpty }) {
+      self.init(outer: i, inner: innerCollection(c[i]).startIndex)
+    }
+    else {
+      self.init(endIn: c)
+    }
+  }
+  
+  internal init<C: Collection>(firstValidIn c: C)
+    where C.Index == OuterIndex, C.Element: Collection, C.Element.Index == InnerIndex
+  {
+    self.init(firstValidIn: c) { $0 }
+  }
```

Comment:
This one is unused.

Reply:
I know; it all needs comments and tests, but this is a convenience for the case where you have a direct collection-of-collections, with no need for projection to get the inner collection.